docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,200 @@
+# ADR 022: Custom BaseApp panic handling
+
+## Changelog
+
+- 2020 Apr 24: Initial Draft
+
+## Context
+
+Current implementation of BaseApp does not allow developers to write custom error handlers in
+[runTx()](https://github.com/cosmos/cosmos-sdk/blob/bad4ca75f58b182f600396ca350ad844c18fc80b/baseapp/baseapp.go#L539)
+method. We think that this method can be more flexible and can give SDK users more options for customizations without
+the need to rewrite whole BaseApp. Also there's one special case for `sdk.ErrorOutOfGas` error handling, that case
+might be handled in a "standard" way (middleware) alongside the others.
+
+We propose middleware-solution, which could help developers implement the following cases:
+* add external logging (let's say sending reports to external services like [Sentry](https://sentry.io));
+* call panic for specific error cases;
+
+It will also make `OutOfGas` case and `default` case one of the middlewares.
+`Default` case wraps recovery object to an error and logs it ([example middleware implementation](#Recovery-middleware)).
+
+## Decision
+
+### Design
+
+#### Overview
+
+Instead of hardcoding custom error handling into BaseApp we suggest using set of middlewares which can be customized
+externally and will allow developers use as many custom error handlers as they want. Implementation with tests
+can be found [here](https://github.com/cosmos/cosmos-sdk/pull/6053).
+
+#### Implementation details
+
+##### Recovery handler
+
+We add a `recover()` object handler type:
+
+```go
+type RecoveryHandler func(recoveryObj interface{}) error
+```
+
+This handler receives an object, processes it and returns contextually enriched `error` or `nil`
+if object is not a handle's target type.
+
+An example:
+
+```go
+func(recoveryObj interface{}) error {
+    err, ok := recoveryObj.(error)
+    if !ok { return nil }
+    
+    if someSpecificError.Is(err) {
+        panic(customPanicMsg)
+    } else {
+        return nil
+    }
+}
+```
+
+This example breaks the application execution, but it also might enrich the error's context like the `OutOfGas` handler.
+
+##### Recovery middleware
+
+We also add a middleware type:
+
+```go
+type recoveryMiddleware func(recoveryObj interface{}) (recoveryMiddleware, error)
+
+func newRecoveryMiddleware(handler RecoveryHandler, next recoveryMiddleware) recoveryMiddleware {
+    return func(recoveryObj interface{}) (recoveryMiddleware, error) {
+        if err := handler(recoveryObj); err != nil {
+            return nil, err
+        }
+        return next, nil
+    }
+}
+```
+
+Function receives a `recover()` object and returns (`next middleware`, `nil`) if object wasn't handled (not a target type)
+or (`nil`, `error`) if input object was handled and other middlewares in the chain should not be executed.
```

Comment:
We should note here that `error` itself can be nil, correct? Or will it always be an error?

Reply:
Description updated:
```
Function receives a `recover()` object and returns:
* (next `recoveryMiddleware`, `nil`) if object wasn't handled (not a target type) by `RecoveryHandler`;
* (`nil`, not nil `error`) if input object was handled and other middlewares in the chain should not be executed;
* (`nil`, `nil`) this is an invalid behaviour 'cause in that case panic recovery might not be properly handled;
This can be avoided by always using a `default` as a rightmost middleware in chain (always returns an `error`'); 
```